From 20b680f189df97e032830c5751621939920d6bdc Mon Sep 17 00:00:00 2001
From: Quallenauge <Hamsi2k@freenet.de>
Date: Thu, 19 Jul 2018 12:26:20 +0200
Subject: [PATCH 4/8] livedisplay: Query active state via native call.

Instead of using a cached member variable which holds the
cached state of the CABL functionality the state is queried
directly via disp_api_is_active_feature_on() function.

Change-Id: Ic256912fcdcf35fabdf51a8eb6c8e842ac5f0ca8
---
 livedisplay/1.0/default/controller/LegacyMMController.cpp    | 5 +++++
 livedisplay/1.0/default/impl/LegacyMM.cpp                    | 5 ++---
 .../1.0/default/include/controller/LegacyMMController.h      | 3 +++
 livedisplay/1.0/default/include/impl/LegacyMM.h              | 1 -
 4 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/livedisplay/1.0/default/controller/LegacyMMController.cpp b/livedisplay/1.0/default/controller/LegacyMMController.cpp
index 0515062..e244dae 100644
--- a/livedisplay/1.0/default/controller/LegacyMMController.cpp
+++ b/livedisplay/1.0/default/controller/LegacyMMController.cpp
@@ -44,6 +44,7 @@
     MACRO(get_pa_config)            \
     MACRO(set_pa_config)            \
     MACRO(active_feature_control)   \
+    MACRO(is_active_feature_on)     \
     MACRO(supported)
 
 #define CONTROLLER_CHECK(function, ...)    \
@@ -153,6 +154,10 @@ int32_t LegacyMMController::active_feature_control(int32_t disp_id, uint32_t fea
     CONTROLLER_CHECK(active_feature_control, disp_id, feature_id, req_id);
 }
 
+int32_t LegacyMMController::is_active_feature_on(int32_t disp_id, uint32_t feature_id) {
+    CONTROLLER_CHECK(is_active_feature_on, disp_id, feature_id);
+}
+
 int32_t LegacyMMController::supported(int32_t disp_id, uint32_t feature_id) {
     CONTROLLER_CHECK(supported, disp_id, feature_id);
 }
diff --git a/livedisplay/1.0/default/impl/LegacyMM.cpp b/livedisplay/1.0/default/impl/LegacyMM.cpp
index 3765c86..76a2123 100644
--- a/livedisplay/1.0/default/impl/LegacyMM.cpp
+++ b/livedisplay/1.0/default/impl/LegacyMM.cpp
@@ -355,18 +355,17 @@ sp<disp_mode> LegacyMM::getDisplayModeById(int32_t id) {
 
 status_t LegacyMM::setAdaptiveBacklightEnabled(bool enabled) {
     status_t rc = NO_INIT;
-    if (enabled == mCachedCABLStatus) {
+    if (enabled == isAdaptiveBacklightEnabled()) {
         return OK;
     }
     if (!mController->active_feature_control(0, 0, !enabled)){
         rc = OK;
-        mCachedCABLStatus = enabled;
     }
     return rc;
 }
 
 bool LegacyMM::isAdaptiveBacklightEnabled() {
-    return mCachedCABLStatus;
+    return mController->is_active_feature_on(0, 0) > 0;
 }
 
 }  // namespace implementation
diff --git a/livedisplay/1.0/default/include/controller/LegacyMMController.h b/livedisplay/1.0/default/include/controller/LegacyMMController.h
index 4185892..5e0a7ec 100644
--- a/livedisplay/1.0/default/include/controller/LegacyMMController.h
+++ b/livedisplay/1.0/default/include/controller/LegacyMMController.h
@@ -45,6 +45,7 @@ class LegacyMMController {
     int32_t get_pa_config(int32_t disp_id, void* cfg);
     int32_t set_pa_config(int32_t disp_id, void* cfg);
     int32_t active_feature_control(int32_t disp_id, uint32_t feature_id, int32_t req_id);
+    int32_t is_active_feature_on(int32_t disp_id, uint32_t feature_id);
     int32_t supported(int32_t disp_id, uint32_t feature_id);
 
   private:
@@ -62,6 +63,7 @@ class LegacyMMController {
     typedef int32_t (*disp_api_get_pa_config)(int32_t, void*);
     typedef int32_t (*disp_api_set_pa_config)(int32_t, void*);
     typedef int32_t (*disp_api_active_feature_control)(int32_t disp_id, uint32_t feature_id, int32_t req_id);
+    typedef int32_t (*disp_api_is_active_feature_on)(int32_t, uint32_t);
     typedef int32_t (*disp_api_supported)(int32_t, uint32_t);
 
     std::shared_ptr<void> mHandle;
@@ -80,6 +82,7 @@ class LegacyMMController {
     disp_api_get_pa_config mFn_get_pa_config;
     disp_api_set_pa_config mFn_set_pa_config;
     disp_api_active_feature_control mFn_active_feature_control;
+    disp_api_is_active_feature_on mFn_is_active_feature_on;
     disp_api_supported mFn_supported;
 };
 
diff --git a/livedisplay/1.0/default/include/impl/LegacyMM.h b/livedisplay/1.0/default/include/impl/LegacyMM.h
index a53d060..6a551ad 100644
--- a/livedisplay/1.0/default/include/impl/LegacyMM.h
+++ b/livedisplay/1.0/default/include/impl/LegacyMM.h
@@ -61,7 +61,6 @@ class LegacyMM : public ColorBackend {
 
   private:
     bool mCABLEnabled;
-    bool mCachedCABLStatus;
     uint32_t getNumDisplayModes();
     android::sp<disp_mode> getDisplayModeById(int32_t id);
 
-- 
2.17.1

